### PR TITLE
Merge from wnprc18.3 svn r. 64002 - 64108

### PR DESCRIPTION
**Pull request by labkey-bpatel**
-Item 5826: [WNPRC] Update DeleteBillingPeriodAction to delete from ehr_billing.invoice, invoiceItems and invoiceRuns with an audit trail.
-Update BillingUtils.isBilling() to make it functional instead of it always returning true.

### CODE REVIEW

File: ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java [hunk 1]
```diff
@@ -52,43 +59,57 @@ public static EHR_BillingManager get()
...
             TableSelector tsMiscCharges2 = new TableSelector(miscCharges, miscChargesFilter, null);
-            ret.add(tsMiscCharges2.getRowCount() + " records from misc charges will be removed from the deleted invoice, which means they will be picked up by the next billing period.  They are not deleted.");
+            ret.add(tsMiscCharges2.getRowCount() + " invoice records from misc charges will be removed from the deleted invoice, which means they will be picked up by the next billing period.  They are not deleted.");
         }
         else
         {
             try (DbScope.Transaction transaction = ExperimentService.get().ensureTransaction())
             {
-                long deleted1 = Table.delete(invoicedItems, invoiceRunFilter);
+                Map<String, Object>[] invoicedItemsRows = tsInvItems.getMapArray();
```
Comment (labkey-jeckels) [inline ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: Suggest using getMapCollection() instead. I know there's plenty of other usages in the code for getMapArray(), but mixing generics and arrays tends to be slightly uglier

Comment (labkey-jeckels) [reply ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: Or better yet, can we avoid doing the select and stick with the filter approach instead?

File: ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java [hunk 2]
```diff
@@ -52,43 +59,57 @@ public static EHR_BillingManager get()
...
+                deleteInvoicedRows(invoicedItems, invoicedItemsRows, user, container);
+
```
Comment (labkey-jeckels) [inline ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: As discussed, recommend just doing an audit entry for the whole run's deletion, don't need separate audit entries for every invoicedItem row

File: ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java [hunk 3]
```diff
@@ -52,43 +59,57 @@ public static EHR_BillingManager get()
         return _instance;
     }
 
-    public List<String> deleteBillingRuns(User user, Collection<String> pks, boolean testOnly)
+    public List<String> deleteBillingRuns(User user, Container container, Collection<String> pks, boolean testOnly) throws SQLException, QueryUpdateServiceException, BatchValidationException, InvalidKeyException
```
[the remark this answers is not in the document]
Comment (labkey-bpatel) [reply ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: getUpdateService().deleteRows() in deleteBillingRuns() throws SQLException, and throwing RuntimeSQLException still requires either catching or throwing SQLException.

deleteRows(...) signature in QueryUpdateService.java: 
List<Map<String,Object>> deleteRows(User user, Container container, List<Map<String, Object>> keys, @Nullable Map<Enum, Object> configParameters, @Nullable Map<String, Object> extraScriptContext)
            **throws InvalidKeyException, BatchValidationException, QueryUpdateServiceException, SQLException**;

File: ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java (hunk 2 above)
Comment (labkey-bpatel) [reply ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: Done.
I've set audit logging to none for invoicedItems and invoice tables.
Also, briefly looked to see if I could change the audit log message to include no. of rows deleted from invoicedItems and invoice - but didn't find anything that was easy to do without creating a custom audit event and provider, unless you have suggestions if we want to go that route.

Comment (labkey-bpatel) [reply ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: This is now updated to use Table.delete()

File: ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java [hunk 4]
```diff
@@ -1,18 +1,3 @@
-/*
```
Comment (labkey-jeckels) [inline ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: Looks like the copyright is disappearing. Can you restore?

File: ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (labkey-jeckels) [reply ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: OK. I'd like to remove SQLException from that throws list, but here I recommend catching it and rethrowing it as a RuntimeSQLException.

File: ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java (hunk 4 above)
Comment (labkey-bpatel) [reply ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: Restored!

File: ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (labkey-bpatel) [reply ehr_billing/src/org/labkey/ehr_billing/EHR_BillingManager.java]: done!

**Pull request merged**